File: Sources/TokenTrackerCore/UsageIssueFormatter.swift
import Foundation

public enum UsageIssueKind: String, Equatable, Sendable {
    case ok
    case disabled
    case rateLimited
    case missingCredentials
    case invalidResponse
    case timedOut
    case network
    case httpStatus
    case usingCachedData
    case unavailable
}

public struct UsageIssue: Equatable, Sendable {
    public let kind: UsageIssueKind
    public let title: String
    public let detail: String
    public let recovery: String?
    public let technicalDetail: String?

    public init(kind: UsageIssueKind, title: String, detail: String, recovery: String?, technicalDetail: String?) {
        self.kind = kind
        self.title = title
        self.detail = detail
        self.recovery = recovery
        self.technicalDetail = technicalDetail
    }
}

public enum UsageIssueFormatter {
    public static func issue(for usage: ProviderUsage, localizer: Localizer = Localizer(language: .english)) -> UsageIssue {
        if usage.source == .staleCache {
            let technicalDetail = usage.error
            return UsageIssue(
                kind: .usingCachedData,
                title: localizer.text(.statusUsingCachedData),
                detail: localizer.text(.statusUsingCachedDataDetail),
                recovery: recovery(for: technicalDetail, localizer: localizer) ?? localizer.text(.recoveryTryAgainLater),
                technicalDetail: technicalDetail
            )
        }

        guard let error = usage.error, !error.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            if usage.isAvailable {
                return UsageIssue(
                    kind: .ok,
                    title: localizer.text(.statusOK),
                    detail: localizer.text(.statusOKDetail),
                    recovery: nil,
                    technicalDetail: nil
                )
            }
            return UsageIssue(
                kind: .unavailable,
                title: localizer.text(.statusUnavailable),
                detail: localizer.text(.statusUnavailableDetail),
                recovery: localizer.text(.recoveryRefreshLater),
                technicalDetail: nil
            )
        }

        let kind = kind(forError: error)
        return UsageIssue(
            kind: kind,
            title: title(for: kind, localizer: localizer),
            detail: detail(for: kind, localizer: localizer),
            recovery: recovery(for: error, localizer: localizer),
            technicalDetail: error
        )
    }

    public static func kind(forError error: String) -> UsageIssueKind {
        let normalized = error.trimmingCharacters(in: .whitespacesAndNewlines)
        let lower = normalized.lowercased()

        if lower == "disabled" {
            return .disabled
        }
        if lower.contains("http 429") || lower.contains("too many requests") {
            return .rateLimited
        }
        if lower.contains("missing credentials") {
            return .missingCredentials
        }
        if lower.contains("invalid response") {
            return .invalidResponse
        }
        if lower.contains("timed out") {
            return .timedOut
        }
        if lower.contains("network error") {
            return .network
        }
        if lower.hasPrefix("http ") {
            return .httpStatus
        }
        return .unavailable
    }

    private static func title(for kind: UsageIssueKind, localizer: Localizer) -> String {
        switch kind {
        case .ok:
            return localizer.text(.statusOK)
        case .disabled:
            return localizer.text(.statusDisabledProvider)
        case .rateLimited:
            return localizer.text(.statusRateLimited)
        case .missingCredentials:
            return localizer.text(.statusMissingCredentials)
        case .invalidResponse:
            return localizer.text(.statusInvalidResponse)
        case .timedOut:
            return localizer.text(.statusTimedOut)
        case .network:
            return localizer.text(.statusNetworkIssue)
        case .httpStatus:
            return localizer.text(.statusHTTPError)
        case .usingCachedData:
            return localizer.text(.statusUsingCachedData)
        case .unavailable:
            return localizer.text(.statusUnavailable)
        }
    }

    private static func detail(for kind: UsageIssueKind, localizer: Localizer) -> String {
        switch kind {
        case .ok:
            return localizer.text(.statusOKDetail)
        case .disabled:
            return localizer.text(.statusDisabledProviderDetail)
        case .rateLimited:
            return localizer.text(.statusRateLimitedDetail)
        case .missingCredentials:
            return localizer.text(.statusMissingCredentialsDetail)
        case .invalidResponse:
            return localizer.text(.statusInvalidResponseDetail)
        case .timedOut:
            return localizer.text(.statusTimedOutDetail)
        case .network:
            return localizer.text(.statusNetworkIssueDetail)
        case .httpStatus:
            return localizer.text(.statusHTTPErrorDetail)
        case .usingCachedData:
            return localizer.text(.statusUsingCachedDataDetail)
        case .unavailable:
            return localizer.text(.statusUnavailableDetail)
        }
    }

    private static func recovery(for error: String?, localizer: Localizer) -> String? {
        guard let error else {
            return nil
        }
        switch kind(forError: error) {
        case .disabled:
            return localizer.text(.recoveryEnableProvider)
        case .rateLimited:
            return localizer.text(.recoveryWaitForCooldown)
        case .missingCredentials:
            return localizer.text(.recoveryCheckCredentials)
        case .invalidResponse:
            return localizer.text(.recoveryUpdateOrTryLater)
        case .timedOut, .network:
            return localizer.text(.recoveryCheckNetwork)
        case .httpStatus, .unavailable:
            return localizer.text(.recoveryRefreshLater)
        case .ok, .usingCachedData:
            return nil
        }
    }
}

File: windows/TokenTracker.Windows.Core/UsageIssueFormatter.cs
namespace TokenTracker.Windows.Core;

public enum UsageIssueKind
{
    Ok,
    Disabled,
    RateLimited,
    MissingCredentials,
    ExpiredCredentials,
    InvalidResponse,
    TimedOut,
    Network,
    HttpStatus,
    UsingCachedData,
    Unavailable
}

public sealed record UsageIssue(
    UsageIssueKind Kind,
    string Title,
    string Detail,
    string? Recovery,
    string? TechnicalDetail);

public static class UsageIssueFormatter
{
    public static UsageIssue Issue(ProviderUsage usage, Localizer? localizer = null)
    {
        localizer ??= new Localizer(AppLanguage.English);
        if (usage.Source == UsageSource.StaleCache)
        {
            var technicalDetail = usage.Error;
            return new UsageIssue(
                UsageIssueKind.UsingCachedData,
                localizer.Text(L10nKey.StatusUsingCachedData),
                localizer.Text(L10nKey.StatusUsingCachedDataDetail),
                Recovery(technicalDetail, usage.Provider, localizer) ?? localizer.Text(L10nKey.RecoveryTryAgainLater),
                technicalDetail);
        }

        if (string.IsNullOrWhiteSpace(usage.Error))
        {
            return usage.IsAvailable
                ? new UsageIssue(
                    UsageIssueKind.Ok,
                    localizer.Text(L10nKey.StatusOk),
                    localizer.Text(L10nKey.StatusOkDetail),
                    null,
                    null)
                : new UsageIssue(
                    UsageIssueKind.Unavailable,
                    localizer.Text(L10nKey.StatusUnavailable),
                    localizer.Text(L10nKey.StatusUnavailableDetail),
                    localizer.Text(L10nKey.RecoveryRefreshLater),
                    null);
        }

        var kind = Kind(usage.Error);
        return new UsageIssue(
            kind,
            Title(kind, localizer),
            Detail(kind, localizer),
            Recovery(usage.Error, usage.Provider, localizer),
            usage.Error);
    }

    public static UsageIssueKind Kind(string error)
    {
        var normalized = error.Trim();
        var lower = normalized.ToLowerInvariant();
        if (lower == "disabled")
        {
            return UsageIssueKind.Disabled;
        }

        if (lower.Contains("http 429", StringComparison.Ordinal) ||
            lower.Contains("too many requests", StringComparison.Ordinal))
        {
            return UsageIssueKind.RateLimited;
        }

        if (lower.Contains("missing credentials", StringComparison.Ordinal))
        {
            return UsageIssueKind.MissingCredentials;
        }

        // 401/403 from either usage endpoint is, in practice, always a lapsed
        // access token: Token Tracker reads whatever the provider CLI last stored
        // and never refreshes it. Reporting it as a generic HTTP error left the
        // user with "refresh again later", which never fixes it.
        if (lower.Contains("credentials expired", StringComparison.Ordinal) ||
            lower.Contains("http 401", StringComparison.Ordinal) ||
            lower.Contains("http 403", StringComparison.Ordinal))
        {
            return UsageIssueKind.ExpiredCredentials;
        }

        if (lower.Contains("invalid response", StringComparison.Ordinal))
        {
            return UsageIssueKind.InvalidResponse;
        }

        if (lower.Contains("timed out", StringComparison.Ordinal))
        {
            return UsageIssueKind.TimedOut;
        }

        if (lower.Contains("network error", StringComparison.Ordinal))
        {
            return UsageIssueKind.Network;
        }

        if (lower.StartsWith("http ", StringComparison.Ordinal))
        {
            return UsageIssueKind.HttpStatus;
        }

        return UsageIssueKind.Unavailable;
    }

    private static string Title(UsageIssueKind kind, Localizer localizer) => kind switch
    {
        UsageIssueKind.Ok => localizer.Text(L10nKey.StatusOk),
        UsageIssueKind.Disabled => localizer.Text(L10nKey.StatusDisabledProvider),
        UsageIssueKind.RateLimited => localizer.Text(L10nKey.StatusRateLimited),
        UsageIssueKind.MissingCredentials => localizer.Text(L10nKey.StatusMissingCredentials),
        UsageIssueKind.ExpiredCredentials => localizer.Text(L10nKey.StatusExpiredCredentials),
        UsageIssueKind.InvalidResponse => localizer.Text(L10nKey.StatusInvalidResponse),
        UsageIssueKind.TimedOut => localizer.Text(L10nKey.StatusTimedOut),
        UsageIssueKind.Network => localizer.Text(L10nKey.StatusNetworkIssue),
        UsageIssueKind.HttpStatus => localizer.Text(L10nKey.StatusHttpError),
        UsageIssueKind.UsingCachedData => localizer.Text(L10nKey.StatusUsingCachedData),
        UsageIssueKind.Unavailable => localizer.Text(L10nKey.StatusUnavailable),
        _ => localizer.Text(L10nKey.StatusUnavailable)
    };

    private static string Detail(UsageIssueKind kind, Localizer localizer) => kind switch
    {
        UsageIssueKind.Ok => localizer.Text(L10nKey.StatusOkDetail),
        UsageIssueKind.Disabled => localizer.Text(L10nKey.StatusDisabledProviderDetail),
        UsageIssueKind.RateLimited => localizer.Text(L10nKey.StatusRateLimitedDetail),
        UsageIssueKind.MissingCredentials => localizer.Text(L10nKey.StatusMissingCredentialsDetail),
        UsageIssueKind.ExpiredCredentials => localizer.Text(L10nKey.StatusExpiredCredentialsDetail),
        UsageIssueKind.InvalidResponse => localizer.Text(L10nKey.StatusInvalidResponseDetail),
        UsageIssueKind.TimedOut => localizer.Text(L10nKey.StatusTimedOutDetail),
        UsageIssueKind.Network => localizer.Text(L10nKey.StatusNetworkIssueDetail),
        UsageIssueKind.HttpStatus => localizer.Text(L10nKey.StatusHttpErrorDetail),
        UsageIssueKind.UsingCachedData => localizer.Text(L10nKey.StatusUsingCachedDataDetail),
        UsageIssueKind.Unavailable => localizer.Text(L10nKey.StatusUnavailableDetail),
        _ => localizer.Text(L10nKey.StatusUnavailableDetail)
    };

    private static string? Recovery(string? error, Provider provider, Localizer localizer)
    {
        if (string.IsNullOrWhiteSpace(error))
        {
            return null;
        }

        return Kind(error) switch
        {
            UsageIssueKind.Disabled => localizer.Text(L10nKey.RecoveryEnableProvider),
            UsageIssueKind.RateLimited => localizer.Text(L10nKey.RecoveryWaitForCooldown),
            UsageIssueKind.MissingCredentials => localizer.Text(L10nKey.RecoveryCheckCredentials),
            // Name only the command the user actually has to run. Listing both
            // providers made them read past the half that did not apply to them.
            UsageIssueKind.ExpiredCredentials => provider == Provider.Claude
                ? localizer.Text(L10nKey.RecoverySignInAgainClaude)
                : localizer.Text(L10nKey.RecoverySignInAgainCodex),
            UsageIssueKind.InvalidResponse => localizer.Text(L10nKey.RecoveryUpdateOrTryLater),
            UsageIssueKind.TimedOut or UsageIssueKind.Network => localizer.Text(L10nKey.RecoveryCheckNetwork),
            UsageIssueKind.HttpStatus or UsageIssueKind.Unavailable => localizer.Text(L10nKey.RecoveryRefreshLater),
            UsageIssueKind.Ok or UsageIssueKind.UsingCachedData => null,
            _ => null
        };
    }
}
